### PR TITLE
Make building man pages optional through option ENABLE_DOC.

### LINKED ISSUE #1481: Explicitly enable or disable building man pages in cmake configuration
## Description

As man pages are built using `asciidoctor`, current implementation just tries to autodetect presence of `asciidoctor` and depending on result either builds manpages or not. Now I am trying to create package for source-based Gentoo distro, and there building documentation is optional and configurable part, depending on user preferences. If user enables `doc` USE-flag either globally or specifically for this package, documentation will be built. And if it is enabled, additional build-time dependency (`asciidoctor` package in this case) should be pulled and installed first. So configuration of rnp build should be more deterministic. I propose to add new cmake variable like `ENABLE_DOC` or `ENABLE_MANPAGES`

## Expected Behavior

If documentation build is explicitly enabled (like -DENABLE_DOC=On), cmake should check for presence of `asciidoctor` and fail to continue if `asciidoctor` is not found. If it is explicitly disabled, then it should not check for `asciidoctor` at all and skip building manpages. Third option (like ENABLE_DOC=Auto) can be implemented and set as a default value in order to keep the current behavior.

## Actual Behavior

`asciidoctor` presence is automatically detected during build and there is no way to disable it or strictly require this dependency for succesfull build

### PR DESCRIPTION
**Pull request by antonsviridenko**
Closes #1481

### CODE REVIEW

**Comment by antonsviridenko** [comment]
Soo, as failed tests show, either asciidoctor should be installed in CI configurations, or building manpages should be disabled.

CMake's `option()` does not support three-state values, I've found possible workaround, not sure if it is suitable:

https://cmake.org/pipermail/cmake/2016-October/064342.html

**Comment by antonsviridenko** [comment]
@ni4 any ideas how to fix failed tests? Where 'asciidoctor' can be added as a dependency exactly?

**Comment by ni4** [comment]
@antonsviridenko Sorry for a late reply.
It could be better to use string variable instead of option, i.e.
```
if (NOT BUILD_DOC)
  set(BUILD_DOC "auto" CACHE STRING
    "Control building of documentation, which requires asciidoctor. Possible values are on, off and auto. Default is auto, which builds documentation only if asciidoctor is found."
    FORCE
  )
endif()
// not sure that one below will work - probably you'll need temporary variable
string(TOLOWER "${BUILD_DOC}" BUILD_DOC)

```

So fail only in case if BUILD_DOC is forcibly set to on.
I would suggest to not fail if asciidoctor is not found, and just issue a warning as it was before.
But, some tests could be added which checks whether man is built correctly under the windows, if asciidoctor is present.

**Comment by ni4** [comment]
Accidentally hit return. For the MSYS test asciidoctor may be installed by adding it to packages in function `msys_install()`, `before_install.sh`.
For the MSVC build i'm not sure, documentation tells about Chocolatey or Rubyinstaller. Maybe @rrrooommmaaa could help here.

**Comment by antonsviridenko** [comment]
My latest commit did not make any improvements :(

ping @rrrooommmaaa

**Comment by ni4** [comment]
@antonsviridenko Yeah, since you still use Off/On for the BUILD_DOC, instead of adding relaxed 'auto' mode.

**Comment by ni4** [comment]
@antonsviridenko Just came across possibly helpful cmake setup: https://stackoverflow.com/questions/39670879/cmake-variable-with-limited-valid-values

Basically, we may have On, Off or Auto value for the `BUILD_DOC` (default auto), and then:
- if it is Auto then build docs only if asciidoctor is available
- if it is On, then try to build docs and report a fatal error if asciidoctor is not found
- if it is Off then just do not build and search for asciidoctor.

**Comment by antonsviridenko** [comment]
yeah, looks interesting

**Review by rrrooommmaaa — APPROVED** [review]
LGTM!

File: CMakeLists.txt [hunk 1]
```diff
@@ -137,7 +143,9 @@ if (ENABLE_SANITIZERS)
 endif()
 
 # adoc for man generation
-include(AdocMan)
+if (ENABLE_DOC)
```
Comment (ni4) [inline CMakeLists.txt:146]: Will this work correctly if ENABLE_DOC is not upper-case, i.e. was set like -DENABLE_DOC=on or -DENABLE_DOC=off ?

File: cmake/Modules/AdocMan.cmake [hunk 2]
```diff
@@ -44,7 +44,13 @@ find_program(ADOCCOMMAND_PATH
 )
 
 if(NOT EXISTS ${ADOCCOMMAND_PATH})
-  message(WARNING "AsciiDoc processor not found, man pages will not be generated. Install asciidoctor or use the CMAKE_PROGRAM_PATH variable.")
+  set(ADOC_MISSING_MSG "AsciiDoc processor not found, man pages will not be generated. Install asciidoctor or use the CMAKE_PROGRAM_PATH variable.")
+
+  if (ENABLE_DOC STREQUAL "Auto")
```
Comment (ni4) [inline cmake/Modules/AdocMan.cmake:49]: Will this work correctly if ENABLE_DOC was set in another case, like -DENABLE_DOC="auto" ?

**Review by ni4 — APPROVED** [review]
LGTM, just few concerns which I cannot check right now.
Feel free to merge once those concerns are excessive :)

File: cmake/Modules/AdocMan.cmake (hunk 2 above)
Comment (antonsviridenko) [reply cmake/Modules/AdocMan.cmake:49]: It appears that no, STREQUAL is case-sensitive.
https://cmake.org/pipermail/cmake/2012-June/050651.html

Meh, things were easier in autotools, why did rnp switch to cmake :D

File: CMakeLists.txt (hunk 1 above)
Comment (antonsviridenko) [reply CMakeLists.txt:146]: Yes, should work correctly


>    True if the constant is 1, ON, YES, TRUE, Y, or a non-zero number. False if the constant is 0, OFF, NO, FALSE, N, IGNORE, NOTFOUND, the empty string, or ends in the suffix -NOTFOUND. **Named boolean constants are case-insensitive**. If the argument is not one of these specific constants, it is treated as a variable or string and the following signature is used.

**Pull request merged**